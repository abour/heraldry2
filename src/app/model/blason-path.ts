export type SimpleBlasonPath = 'field' | 'ordinary' | 'charge';
export type CompoundedBlasonPath = [number, SimpleBlasonPath | null];
export type BlasonPath = SimpleBlasonPath | CompoundedBlasonPath;
