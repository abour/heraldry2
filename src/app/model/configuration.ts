import { TinctureConfiguration } from './tincture-configuration';

export const shieldShapes = ['heater', 'spanish'  /*, 'french', 'swiss', */] as const;
export type ShieldShape = typeof shieldShapes[number];

export type Configuration = {
  shieldShape: ShieldShape;
  tinctureConfiguration: TinctureConfiguration;
};
