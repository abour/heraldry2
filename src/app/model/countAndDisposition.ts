export const numberToNameMap = {
  2: 'two',
  3: 'three',
  4: 'four',
  5: 'five',
  6: 'six',
  7: 'seven',
  8: 'eight',
  9: 'nine',
  10: 'ten',
  11: 'eleven',
  12: 'twelve',
  13: 'thirteen',
  14: 'fourteen',
  15: 'fifteen',
  16: 'sixteen',
  17: 'seventeen',
  18: 'eighteen',
  19: 'nineteen',
  20: 'twenty',
} as const;
export type StringifiableNumber = keyof typeof numberToNameMap;
export function stringifyNumber(n: StringifiableNumber): string {
  return numberToNameMap[n];
}

export type SupportedNumber = 1 | StringifiableNumber;
export const supportedNumbers: ReadonlyArray<SupportedNumber> = [
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
  20,
] as const;
export function isNotOne(i: SupportedNumber): i is Exclude<SupportedNumber, 1> {
  return i !== 1;
}
export const availableDispositions = ['default', 'pale', 'fess', 'bend', 'bendSinister'] as const;
export type Disposition = typeof availableDispositions[number];
export type CountAndDisposition = { count: SupportedNumber; disposition: Disposition };
