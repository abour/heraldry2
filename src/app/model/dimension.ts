export type Dimension = { width: number; height: number };

export function scale({ width, height }: Dimension, scaleFactor: number): Dimension {
  return {
    width: width * scaleFactor,
    height: height * scaleFactor,
  };
}
