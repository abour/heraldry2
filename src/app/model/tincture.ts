export const argent = { tricking: ['ar.'], name: 'argent' } as const;
export type Argent = typeof argent;
export const or = { tricking: ['o.'], name: 'or' } as const;
export type Or = typeof or;

export const gules = { tricking: ['gu.'], name: 'gules' } as const;
export type Gules = typeof gules;
export const sable = { tricking: ['s.', 'sa.'], name: 'sable' } as const;
export type Sable = typeof sable;
export const azure = { tricking: ['as.', 'bl.', 'b.'], name: 'azure' } as const;
export type Azure = typeof azure;
export const vert = { tricking: ['vt.', 'v.'], name: 'vert' } as const;
export type Vert = typeof vert;
export const purpure = { tricking: ['purp.', 'pu.', 'p.'], name: 'purpure' } as const;
export type Purpure = typeof purpure;

export const murrey = { tricking: 'm.', name: 'murrey' } as const;
export type Murrey = typeof murrey;
export const sanguine = { tricking: [], name: 'sanguine' } as const;
export type Sanguine = typeof sanguine;
export const tenne = { tricking: [], name: 'tenne' } as const;
export type Tenne = typeof tenne;

export type Metal = Argent | Or;
export type Colours = Gules | Sable | Azure | Vert | Purpure;
export type Stains = Murrey | Sanguine | Tenne;

export type TinctureName = (Metal | Colours | Stains)['name'];

export const ermine = { name: 'ermine', field: argent, spot: sable } as const;
export type Ermine = typeof ermine;
export const counterErmine = { name: 'counter-ermine', field: sable, spot: argent } as const;
export type CounterErmine = typeof counterErmine;
export const erminois = { name: 'erminois', field: or, spot: sable } as const;
export type Erminois = typeof erminois;
export const pean = { name: 'pean', field: sable, spot: or } as const;
export type Pean = typeof pean;
export const vair = { name: 'vair', field: argent, bell: azure } as const;
export type Vair = typeof vair;
export const counterVair = { name: 'counter-vair', field: argent, bell: azure } as const;
export type CounterVair = typeof counterVair;
export const vairEnPale = { name: 'vair-en-pale', field: argent, bell: azure } as const;
export type VairEnPale = typeof vairEnPale;
export const vairEnPoint = { name: 'vair-en-point', field: argent, bell: azure } as const;
export type VairEnPoint = typeof vairEnPoint;

export const potent = { name: 'potent', field: argent, bell: azure } as const;
export type Potent = typeof potent;
export const counterPotent = { name: 'counter-potent', field: argent, bell: azure } as const;
export type CounterPotent = typeof counterPotent;
export const potentEnPale = { name: 'potent-en-pale', field: argent, bell: azure } as const;
export type PotentEnPale = typeof potentEnPale;
export const potentEnPoint = { name: 'potent-en-point', field: argent, bell: azure } as const;
export type PotentEnPoint = typeof potentEnPoint;

export type Ermines = Ermine | CounterErmine | Erminois | Pean;
export const ermines: ReadonlyArray<Ermines> = [ermine, counterErmine, erminois, pean] as const;

export function isErmine(t: Tincture): t is Ermines {
  return ermines.some((e) => e.name === t.name);
}

export type Vairs = Vair | CounterVair | VairEnPale | VairEnPoint;
export const vairs: ReadonlyArray<Vairs> = [vair, counterVair, vairEnPale, vairEnPoint] as const;
export function isVair(t: Tincture): t is Vairs {
  return vairs.some((e) => e.name === t.name);
}

export type Potents = Potent | CounterPotent | PotentEnPale | PotentEnPoint;
export const potents: ReadonlyArray<Potents> = [potent, counterPotent, potentEnPale, potentEnPoint] as const;
export function isPotent(t: Tincture): t is Potents {
  return potents.some((e) => e.name === t.name);
}

export type Furs = Ermines | Vairs | Potents;

export type MetalsAndColours = Metal | Colours;
export type Tincture = MetalsAndColours | Stains | Furs;

export function isFur(tincture: Tincture): tincture is Furs {
  return isErmine(tincture) || isVair(tincture) || isPotent(tincture);
}

export function isMetal(tincture: Tincture): tincture is Metal {
  return tincture.name === 'argent' || tincture.name === 'or';
}

export const metalAndColours: Array<MetalsAndColours> = [argent, or, gules, sable, azure, vert, purpure];

export const tinctures: Array<Tincture> = [
  ...metalAndColours,
  murrey,
  sanguine,
  tenne,
  ...ermines,
  ...vairs,
  ...potents,
];
