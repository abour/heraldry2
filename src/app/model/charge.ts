import { Tincture } from './tincture';
import { CountAndDisposition } from './countAndDisposition';

export const lionAttitudes = [
  'rampant',
  'passant',
  'statant',
  'salient',
  'sejant',
  'sejant-erect',
  'couchant',
  'dormant',
] as const;
export const lionHeads = ['guardant', 'regardant'] as const;
export const lionTails = ['coward', 'forked', 'crossed', 'nowed'] as const;

export type LionAttitude = typeof lionAttitudes[number];
export type LionHead = typeof lionHeads[number];
export type LionTail = typeof lionTails[number];

export type Lion = {
  name: 'lion';
  attitude: LionAttitude;
  head: LionHead | null;
  tail: LionTail | null;

  tincture: Tincture;
  armedAndLangued: Tincture;
  countAndDisposition: CountAndDisposition;
};

export const eagleAttitudes = ['displayed'] as const;
export type EagleAttitude = typeof eagleAttitudes[number];
export type Eagle = {
  name: 'eagle';
  attitude: EagleAttitude;
  tincture: Tincture;
  beakedAndArmed: Tincture;
  countAndDisposition: CountAndDisposition;
};

export type FleurDeLys = {
  name: 'fleurdelys';
  countAndDisposition: CountAndDisposition;
  tincture: Tincture;
};

export const roundelInsides = ['nothing', 'voided'] as const;
export type RoundelInside = typeof roundelInsides[number];
export type Roundel = {
  name: 'roundel';
  countAndDisposition: CountAndDisposition;
  tincture: Tincture;
  inside: RoundelInside;
};

export const lozengeInsides = ['nothing', 'voided', 'pierced'] as const;
export type LozengeInside = typeof lozengeInsides[number];
export type Lozenge = {
  name: 'lozenge';
  countAndDisposition: CountAndDisposition;
  tincture: Tincture;
  inside: LozengeInside;
};

export const crossLimbs = [
  'hummetty',
  'pattée',
  'potent',
  'cercelée',
  'moline',
  'bottony',
  'crosselet',
  'maltese',
  'flory',
] as const;
export type CrossLimbs = typeof crossLimbs[number];
export type Cross = {
  name: 'cross';
  countAndDisposition: CountAndDisposition;
  tincture: Tincture;
  limbs: CrossLimbs;
};

export type Charge = Lion | Eagle | FleurDeLys | Roundel | Lozenge | Cross;

export const charges: Array<Charge['name']> = ['lion', 'eagle', 'fleurdelys', 'roundel', 'lozenge', 'cross'];
