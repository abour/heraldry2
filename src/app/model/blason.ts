import { Ordinary } from './ordinary';
import { Charge } from './charge';
import { Field } from './field';

export type SimpleBlason = {
  kind: 'simple',
  field: Field;
  ordinary?: Ordinary;
  charge?: Charge;
};
export type QuarterlyBlason = {
  kind: 'quarterly',
  blasons: readonly [SimpleBlason,SimpleBlason,SimpleBlason,SimpleBlason]
}
export type Blason = SimpleBlason | QuarterlyBlason;
