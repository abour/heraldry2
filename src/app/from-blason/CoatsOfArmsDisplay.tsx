import * as React from 'react';
import { Blason } from '../model/blason';
import { isThereFur } from './blason.helpers';
import { uuid } from '../../utils/uuid';
import { ermines, Furs, isFur, potents, Tincture, vairs } from '../model/tincture';
import { Dimension } from '../model/dimension';
import { ErminePatternDef } from './coats-of-arms-parts/ErminePatternDef';
import { VairPatternDef } from './coats-of-arms-parts/VairPatternDef';
import { PotentPatternDef } from './coats-of-arms-parts/PotentPatternDef';
import { Configuration } from '../model/configuration';
import { EscutcheonDisplay } from './coats-of-arms-parts/escutcheon/EscutcheonDisplay';
import { BlasonDisplay } from './coats-of-arms-parts/BlasonDisplay';
import { BlasonPath } from '../model/blason-path';

type Props = {
  blason: Blason;
  dimension: Dimension;
  configuration: Configuration;
  selectBlasonPart: (path: BlasonPath) => void;
};
export function CoatsOfArmsDisplay(props: Props) {
  const configuration = props.configuration;
  const tinctureConfiguration = configuration.tinctureConfiguration;
  const dimension = props.dimension;
  const { width, height } = dimension;

  const patternIds: { [K in Furs['name']]: string } = {
    vair: uuid(),
    'vair-en-pale': uuid(),
    'counter-vair': uuid(),
    ermine: uuid(),
    'counter-ermine': uuid(),
    erminois: uuid(),
    pean: uuid(),
    potent: uuid(),
    'counter-potent': uuid(),
    'potent-en-pale': uuid(),
    'potent-en-point': uuid(),
    'vair-en-point': uuid(),
  };

  function furPatternId(fur: Furs): string {
    return `field-pattern-${patternIds[fur.name]}`;
  }

  function fillFromTincture(tincture: Tincture): string {
    if (isFur(tincture)) {
      return `url(#${furPatternId(tincture)})`;
    } else {
      return tinctureConfiguration[tincture.name];
    }
  }

  const blason = props.blason;

  return (
    <svg width={width} height={height} viewBox={`0 0 ${width} ${height}`} className="coats-of-arms-display">
      <defs>
        {ermines.map((ermine, i) => {
          return isThereFur(blason, ermine.name) ? (
            <ErminePatternDef
              key={ermine.name + i}
              ermine={ermine}
              dimension={dimension}
              patternId={furPatternId(ermine)}
              tinctureConfiguration={tinctureConfiguration}
            />
          ) : (
            ''
          );
        })}
        {vairs.map((vair, i) => {
          return isThereFur(blason, vair.name) ? (
            <VairPatternDef
              key={vair.name + i}
              vair={vair}
              patternId={furPatternId(vair)}
              dimension={dimension}
              tinctureConfiguration={tinctureConfiguration}
            />
          ) : (
            ''
          );
        })}
        {potents.map((potent, i) => {
          return isThereFur(blason, potent.name) ? (
            <PotentPatternDef
              key={potent.name + i}
              potent={potent}
              patternId={furPatternId(potent)}
              dimension={dimension}
              tinctureConfiguration={tinctureConfiguration}
            />
          ) : (
            ''
          );
        })}

        <clipPath id="plain-field-clip-path">
          <EscutcheonDisplay dimension={dimension} shieldShape={configuration.shieldShape} />
        </clipPath>
      </defs>

      <BlasonDisplay
        blason={blason}
        dimension={dimension}
        fillFromTincture={fillFromTincture}
        clipPathId="plain-field-clip-path"
        shieldShape={configuration.shieldShape}
        selectBlasonPart={props.selectBlasonPart}
      />
      <EscutcheonDisplay dimension={dimension} shieldShape={configuration.shieldShape} />
    </svg>
  );
}
