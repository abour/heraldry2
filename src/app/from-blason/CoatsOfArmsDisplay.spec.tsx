import renderer from 'react-test-renderer';
import React from 'react';
import { CoatsOfArmsDisplay } from './CoatsOfArmsDisplay';
import { Blason } from '../model/blason';
import { azure, ermine, gules, purpure, vair } from '../model/tincture';
jest.mock('../../utils/uuid');

describe('CoatsOfArms', () => {
  it('should render a plain blason', () => {
    const blason: Blason = { field: { kind: 'plain', tincture: gules } };
    const component = renderer.create(<CoatsOfArmsDisplay blason={blason} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render a fur blason', () => {
    const blason: Blason = { field: { kind: 'plain', tincture: ermine } };
    const component = renderer.create(<CoatsOfArmsDisplay blason={blason} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render a party blason', () => {
    const blason: Blason = { field: { kind: 'party', per: { name: 'bend', tinctures: [vair, azure] } } };
    const component = renderer.create(<CoatsOfArmsDisplay blason={blason} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render a party blason with some ordinary', () => {
    const blason: Blason = {
      field: { kind: 'party', per: { name: 'bend', tinctures: [vair, azure] } },
      ordinary: { name: 'chief', tincture: purpure },
    };
    const component = renderer.create(<CoatsOfArmsDisplay blason={blason} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
