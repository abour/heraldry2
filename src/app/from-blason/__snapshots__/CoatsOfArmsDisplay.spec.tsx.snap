// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoatsOfArms should render a fur blason 1`] = `
<svg
  height={266.6666666666667}
  viewBox="0 0 200 266.6666666666667"
  width={200}
>
  <defs>
    <symbol
      id="ermine"
      viewBox="0 0 200 240"
    >
      <rect
        fill="white"
        height="100%"
        width="100%"
      />
      <g
        transform="translate(100, 240)"
      >
        <path
          d="M0,-174 C -15,-30 -29,-35 -55,0 L -15,-28 0,0 15,-28 55,0 C 28,-35 15,-30 0,-174 Z"
        />
        <circle
          cy={-226}
          r="14"
        />
        <circle
          cx={-32}
          cy={-194}
          r={14}
        />
        <circle
          cx={32}
          cy={-194}
          r={14}
        />
      </g>
    </symbol>
    <pattern
      height={64.9}
      id="field-pattern-3"
      patternUnits="userSpaceOnUse"
      viewBox="0 0 70 125"
      width={36.36}
    >
      <rect
        fill="white"
        height="100%"
        width="100%"
      />
      <use
        height={60}
        href="#ermine"
        width={35}
        x="0"
        y="0"
      />
      <use
        height={60}
        href="#ermine"
        width={35}
        x="35"
        y="65"
      />
    </pattern>
    <clipPath
      id="plain-field-clip-path"
    >
      <path
        d="M 0 0 H200 V88.8888888888889 A200 200 90 0 1 100 266.6666666666667 A200 200 -90 0 1 0 88.8888888888889 Z"
        fill="transparent"
        stroke="#333"
      />
    </clipPath>
  </defs>
  <g
    clipPath="url(#plain-field-clip-path)"
  >
    <rect
      fill="url(#field-pattern-3)"
      height={266.6666666666667}
      stroke="#333"
      width={200}
      x={0}
      y={0}
    />
  </g>
  <path
    d="M 0 0 H200 V88.8888888888889 A200 200 90 0 1 100 266.6666666666667 A200 200 -90 0 1 0 88.8888888888889 Z"
    fill="transparent"
    stroke="#333"
  />
</svg>
`;

exports[`CoatsOfArms should render a party blason 1`] = `
<svg
  height={266.6666666666667}
  viewBox="0 0 200 266.6666666666667"
  width={200}
>
  <defs>
    <symbol
      id="vair"
      viewBox="0 0 200 200"
    >
      <rect
        fill="white"
        height="100%"
        width="100%"
      />
      <path
        d="M0 200 L200 200 L150 150 L150 50 L100 0 L50 50 L50 150 Z"
        fill="#0000FF"
      />
    </symbol>
    <pattern
      height={80}
      id="field-pattern-6"
      patternUnits="userSpaceOnUse"
      viewBox="0 0 100 200"
      width={40}
    >
      <rect
        fill="white"
        height="100%"
        width="100%"
      />
      <use
        height={100}
        href="#vair"
        width={100}
        x="0"
        y="0"
      />
      <use
        height={100}
        href="#vair"
        width={100}
        x="-50"
        y="100"
      />
      <use
        height={100}
        href="#vair"
        width={100}
        x="50"
        y="100"
      />
    </pattern>
    <clipPath
      id="plain-field-clip-path"
    >
      <path
        d="M 0 0 H200 V88.8888888888889 A200 200 90 0 1 100 266.6666666666667 A200 200 -90 0 1 0 88.8888888888889 Z"
        fill="transparent"
        stroke="#333"
      />
    </clipPath>
  </defs>
  <g
    clipPath="url(#plain-field-clip-path)"
  >
    <g>
      <path
        d="M 0 0 H 200 V 266.6666666666667 Z"
        fill="url(#field-pattern-6)"
        stroke="#333"
      />
      <path
        d="M 0 0 V 266.6666666666667 H 200  Z"
        fill="#0000FF"
        stroke="#333"
      />
    </g>
  </g>
  <path
    d="M 0 0 H200 V88.8888888888889 A200 200 90 0 1 100 266.6666666666667 A200 200 -90 0 1 0 88.8888888888889 Z"
    fill="transparent"
    stroke="#333"
  />
</svg>
`;

exports[`CoatsOfArms should render a party blason with some ordinary 1`] = `
<svg
  height={266.6666666666667}
  viewBox="0 0 200 266.6666666666667"
  width={200}
>
  <defs>
    <symbol
      id="vair"
      viewBox="0 0 200 200"
    >
      <rect
        fill="white"
        height="100%"
        width="100%"
      />
      <path
        d="M0 200 L200 200 L150 150 L150 50 L100 0 L50 50 L50 150 Z"
        fill="#0000FF"
      />
    </symbol>
    <pattern
      height={80}
      id="field-pattern-8"
      patternUnits="userSpaceOnUse"
      viewBox="0 0 100 200"
      width={40}
    >
      <rect
        fill="white"
        height="100%"
        width="100%"
      />
      <use
        height={100}
        href="#vair"
        width={100}
        x="0"
        y="0"
      />
      <use
        height={100}
        href="#vair"
        width={100}
        x="-50"
        y="100"
      />
      <use
        height={100}
        href="#vair"
        width={100}
        x="50"
        y="100"
      />
    </pattern>
    <clipPath
      id="plain-field-clip-path"
    >
      <path
        d="M 0 0 H200 V88.8888888888889 A200 200 90 0 1 100 266.6666666666667 A200 200 -90 0 1 0 88.8888888888889 Z"
        fill="transparent"
        stroke="#333"
      />
    </clipPath>
  </defs>
  <g
    clipPath="url(#plain-field-clip-path)"
  >
    <g>
      <path
        d="M 0 0 H 200 V 266.6666666666667 Z"
        fill="url(#field-pattern-8)"
        stroke="#333"
      />
      <path
        d="M 0 0 V 266.6666666666667 H 200  Z"
        fill="#0000FF"
        stroke="#333"
      />
    </g>
  </g>
  <g
    clipPath="url(#plain-field-clip-path)"
  >
    <rect
      fill="#800080"
      height={53.333333333333336}
      stroke="#333"
      width={200}
      x={0}
      y={0}
    />
  </g>
  <path
    d="M 0 0 H200 V88.8888888888889 A200 200 90 0 1 100 266.6666666666667 A200 200 -90 0 1 0 88.8888888888889 Z"
    fill="transparent"
    stroke="#333"
  />
</svg>
`;

exports[`CoatsOfArms should render a plain blason 1`] = `
<svg
  height={266.6666666666667}
  viewBox="0 0 200 266.6666666666667"
  width={200}
>
  <defs>
    <clipPath
      id="plain-field-clip-path"
    >
      <path
        d="M 0 0 H200 V88.8888888888889 A200 200 90 0 1 100 266.6666666666667 A200 200 -90 0 1 0 88.8888888888889 Z"
        fill="transparent"
        stroke="#333"
      />
    </clipPath>
  </defs>
  <g
    clipPath="url(#plain-field-clip-path)"
  >
    <rect
      fill="#FF0000"
      height={266.6666666666667}
      stroke="#333"
      width={200}
      x={0}
      y={0}
    />
  </g>
  <path
    d="M 0 0 H200 V88.8888888888889 A200 200 90 0 1 100 266.6666666666667 A200 200 -90 0 1 0 88.8888888888889 Z"
    fill="transparent"
    stroke="#333"
  />
</svg>
`;
